// @sqds/multisig/lib/instructions/batchAddTransaction.js
import { createBatchAddTransactionInstruction } from "../generated";
import { getBatchTransactionPda, getProposalPda, getTransactionPda, getVaultPda, } from "../pda";
import { transactionMessageToMultisigTransactionMessageBytes } from "../utils";
export function batchAddTransaction({ vaultIndex, multisigPda, member, rentPayer, batchIndex, transactionIndex, ephemeralSigners, transactionMessage, addressLookupTableAccounts, }) {
    const [proposalPda] = getProposalPda({
        multisigPda,
        transactionIndex: batchIndex,
    });
    const [batchPda] = getTransactionPda({
        multisigPda,
        index: batchIndex,
    });
    const [batchTransactionPda] = getBatchTransactionPda({
        multisigPda,
        batchIndex,
        transactionIndex,
    });
    const [vaultPda] = getVaultPda({
        multisigPda,
        index: vaultIndex,
    });
    const transactionMessageBytes = transactionMessageToMultisigTransactionMessageBytes({
        message: transactionMessage,
        addressLookupTableAccounts,
        vaultPda,
    });
    return createBatchAddTransactionInstruction({
        multisig: multisigPda,
        member,
        proposal: proposalPda,
        rentPayer: rentPayer ?? member,
        batch: batchPda,
        transaction: batchTransactionPda,
    }, {
        args: {
            ephemeralSigners,
            transactionMessage: transactionMessageBytes,
        },
    });
}
