// @sqds/multisig/lib/instructions/vaultTransactionCreate.js
import { createVaultTransactionCreateInstruction } from "../generated";
import { getTransactionPda, getVaultPda } from "../pda";
import { transactionMessageToMultisigTransactionMessageBytes } from "../utils";
export function vaultTransactionCreate({ multisigPda, transactionIndex, creator, rentPayer, vaultIndex, ephemeralSigners, transactionMessage, addressLookupTableAccounts, memo, }) {
    const [vaultPda] = getVaultPda({
        multisigPda,
        index: vaultIndex,
    });
    const [transactionPda] = getTransactionPda({
        multisigPda,
        index: transactionIndex,
    });
    const transactionMessageBytes = transactionMessageToMultisigTransactionMessageBytes({
        message: transactionMessage,
        addressLookupTableAccounts,
        vaultPda,
    });
    return createVaultTransactionCreateInstruction({
        multisig: multisigPda,
        transaction: transactionPda,
        creator,
        rentPayer: rentPayer ?? creator,
    }, {
        args: {
            vaultIndex,
            ephemeralSigners,
            transactionMessage: transactionMessageBytes,
            memo: memo ?? null,
        },
    });
}
