// @sqds/multisig/lib/generated/instructions/multisigRemoveSpendingLimit.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { multisigRemoveSpendingLimitArgsBeet, } from '../types/MultisigRemoveSpendingLimitArgs';
/**
 * @category Instructions
 * @category MultisigRemoveSpendingLimit
 * @category generated
 */
export const multisigRemoveSpendingLimitStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['args', multisigRemoveSpendingLimitArgsBeet],
], 'MultisigRemoveSpendingLimitInstructionArgs');
export const multisigRemoveSpendingLimitInstructionDiscriminator = [
    228, 198, 136, 111, 123, 4, 178, 113,
];
/**
 * Creates a _MultisigRemoveSpendingLimit_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category MultisigRemoveSpendingLimit
 * @category generated
 */
export function createMultisigRemoveSpendingLimitInstruction(accounts, args, programId = new web3.PublicKey('SQDS4ep65T869zMMBKyuUq6aD6EgTu8psMjkvj52pCf')) {
    const [data] = multisigRemoveSpendingLimitStruct.serialize({
        instructionDiscriminator: multisigRemoveSpendingLimitInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.multisig,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.configAuthority,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.spendingLimit,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.rentCollector,
            isWritable: true,
            isSigner: false,
        },
    ];
    if (accounts.anchorRemainingAccounts != null) {
        for (const acc of accounts.anchorRemainingAccounts) {
            keys.push(acc);
        }
    }
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
