// @sqds/multisig/lib/generated/instructions/multisigCreate.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { multisigCreateArgsBeet, } from '../types/MultisigCreateArgs';
/**
 * @category Instructions
 * @category MultisigCreate
 * @category generated
 */
export const multisigCreateStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['args', multisigCreateArgsBeet],
], 'MultisigCreateInstructionArgs');
export const multisigCreateInstructionDiscriminator = [
    122, 77, 80, 159, 84, 88, 90, 197,
];
/**
 * Creates a _MultisigCreate_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category MultisigCreate
 * @category generated
 */
export function createMultisigCreateInstruction(accounts, args, programId = new web3.PublicKey('SQDS4ep65T869zMMBKyuUq6aD6EgTu8psMjkvj52pCf')) {
    const [data] = multisigCreateStruct.serialize({
        instructionDiscriminator: multisigCreateInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.multisig,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.createKey,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.creator,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    if (accounts.anchorRemainingAccounts != null) {
        for (const acc of accounts.anchorRemainingAccounts) {
            keys.push(acc);
        }
    }
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
