// @sqds/multisig/lib/generated/instructions/vaultTransactionCreate.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { vaultTransactionCreateArgsBeet, } from '../types/VaultTransactionCreateArgs';
/**
 * @category Instructions
 * @category VaultTransactionCreate
 * @category generated
 */
export const vaultTransactionCreateStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['args', vaultTransactionCreateArgsBeet],
], 'VaultTransactionCreateInstructionArgs');
export const vaultTransactionCreateInstructionDiscriminator = [
    48, 250, 78, 168, 208, 226, 218, 211,
];
/**
 * Creates a _VaultTransactionCreate_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category VaultTransactionCreate
 * @category generated
 */
export function createVaultTransactionCreateInstruction(accounts, args, programId = new web3.PublicKey('SQDS4ep65T869zMMBKyuUq6aD6EgTu8psMjkvj52pCf')) {
    const [data] = vaultTransactionCreateStruct.serialize({
        instructionDiscriminator: vaultTransactionCreateInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.multisig,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.transaction,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.creator,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.rentPayer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    if (accounts.anchorRemainingAccounts != null) {
        for (const acc of accounts.anchorRemainingAccounts) {
            keys.push(acc);
        }
    }
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
